### PR TITLE
Enable readOnlyRootFilesystem for all Kubevirt containers

### PR DESCRIPTION
**Pull request by davmarro**
#### Before this PR:
KubeVirt containers (`virt-operator`, `virt-api`, `virt-controller`, `virt-exportproxy`) run without `readOnlyRootFilesystem: true`, triggering compliance scan findings.

#### After this PR:
All KubeVirt infrastructure containers have `readOnlyRootFilesystem: true` set in their SecurityContext. An emptyDir volume (`tmp-dir`) is added to `virt-api` only, because it calls `os.MkdirTemp("", ...)` to create a temporary certificates directory.

### Why we need it and why it was done in this way
The following tradeoffs were made:
- A minimalist approach was chosen: emptyDir `/tmp` is added only to `virt-api`, where a confirmed write exists. Other containers don't write to `/tmp` in production code.
- `virt-controller`, `virt-operator`, `virt-exportproxy`, and `virt-synchronization-controller` do not need `/tmp` — no writes to the root filesystem were identified.
- `virt-handler` and `virt-launcher` init container are intentionally excluded: they run with `privileged: true` ([daemonsets.go#L248](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-operator/resource/generate/components/daemonsets.go#L248), [daemonsets.go#L166](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-operator/resource/generate/components/daemonsets.go#L166)). Setting `readOnlyRootFilesystem` on privileged containers provides no security benefit (a privileged container can remount the filesystem rw), while it breaks runtime functionality (`virtqemud` fails to write its pidfile).

The following alternatives were considered:
- Adding `readOnlyRootFilesystem` to all containers including privileged ones — rejected because it causes `virtqemud` pidfile write failures, preventing virt-handler rollout.

Links to places where the discussion took place: 

### Special notes for your reviewer
- `virt-synchronization-controller` is also covered since it uses the same SecurityContext pattern.
- Unit tests verify RORFS on all deployments and DaemonSet containers, plus presence/absence of `tmp-dir` volume.
- E2e tests verify the same properties on deployments in the cluster.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [VEP (virtualization enhancement proposal)](https://github.com/kubevirt/enhancements/blob/main/README.md) was considered and is present (link) or not required
- [X] PR: The PR description is expressive enough and will help future contributors
- [X] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [X] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [X] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [X] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [X] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered
- [X] AI Contributions: The PR abides by the [KubeVirt AI Contribution Policy](https://github.com/kubevirt/community/blob/main/ai-contribution-policy.md).

### Release note

```release-note
Enable readOnlyRootFilesystem for all KubeVirt infrastructure containers
```

### CODE REVIEW

File: tests/operator/operator.go [hunk 1]
```diff
@@ -2543,6 +2543,86 @@ var _ = Describe("[sig-operator]Operator", Serial, decorators.SigOperator, func(
 			verifyAggregateLabels(kubevirt.Client(), "true")
 		})
 	})
+
+	Describe("Security", func() {
+		It("all infrastructure deployments should have readOnlyRootFilesystem set to true", func() {
+			deploymentNames := []string{components.VirtAPIName, components.VirtControllerName, components.VirtOperatorName, components.VirtExportProxyName}
```
[the remark this answers is not in the document]
Comment (davmarro) [reply tests/operator/operator.go]: Fixed, thanks

**Comment by dasionov** [comment]
Hi @davmarro Thank you for this PR :+1: 
Really nice work on this, thanks for pushing it forward 🙏 The control-plane side looks clean and the test coverage is solid.

One thing for a possible follow-up - I think there are still a few containers that aren't covered yet and could take readOnlyRootFilesystem too, mostly node-side:

1. `virt-launcher-image-holder` in the virt-handler daemonset (only added when `imagePullSecrets` are set — it just runs `sleep infinity`, so RORFS is basically free here). Small heads up: it also has no `SecurityContext` at all, so the new handler e2e test that asserts non-nil for every container will fail when `imagePullSecrets` are configured.
2. the `container-disk` containers in the per-VM virt-launcher pods (they only copy into shared emptyDir mounts).
3. the hotplug attachment / trigger pods, and the `virtiofs` containers.

These are all node-scoped so no need to grow this PR - we're planning to pick them up separately. 

/lgtm

**Comment by davmarro** [comment]
> Hi @davmarro Thank you for this PR 👍 Really nice work on this, thanks for pushing it forward 🙏 The control-plane side looks clean and the test coverage is solid.
> 
> One thing for a possible follow-up - I think there are still a few containers that aren't covered yet and could take readOnlyRootFilesystem too, mostly node-side:
> 
> 1. `virt-launcher-image-holder` in the virt-handler daemonset (only added when `imagePullSecrets` are set — it just runs `sleep infinity`, so RORFS is basically free here). Small heads up: it also has no `SecurityContext` at all, so the new handler e2e test that asserts non-nil for every container will fail when `imagePullSecrets` are configured.
> 2. the `container-disk` containers in the per-VM virt-launcher pods (they only copy into shared emptyDir mounts).
> 3. the hotplug attachment / trigger pods, and the `virtiofs` containers.
> 
> These are all node-scoped so no need to grow this PR - we're planning to pick them up separately.
> 
> /lgtm

Hi @dasionov thanks for your feedback. 
I'm going to add a basic security context to the `virt-launcher-image-holder` in this PR while I'll follow with another PR covering the other points mentioned. Thanks!

**Comment by davmarro** [comment]
Hi @dasionov, here's the [change](https://github.com/kubevirt/kubevirt/pull/18853/commits/720e7f349e8a800e38a661a971ed8a087ab641a7) with the new security context for the `virt-launcher-image-holder`

**Comment by dasionov** [comment]
@davmarro Thank You!
/lgtm

**Comment by dasionov** [comment]
/cc @0xFelix 
PTAL :pray:

**Comment by davmarro** [comment]
Hi @0xFelix, any update on the code review ? 
Thanks

File: pkg/virt-operator/resource/generate/components/daemonsets.go [hunk 2]
```diff
@@ -190,6 +191,15 @@ func NewHandlerDaemonSet(config *operatorutil.KubeVirtDeploymentConfig, productN
 			ImagePullPolicy: config.GetImagePullPolicy(),
 			Command:         []string{"/bin/sh", "-c"},
 			Args:            []string{"sleep infinity"},
+			SecurityContext: &corev1.SecurityContext{
+				AllowPrivilegeEscalation: pointer.P(false),
```
Comment (0xFelix) [inline pkg/virt-operator/resource/generate/components/daemonsets.go]: Can you use `new` instead of `pointer.P`?

File: pkg/virt-operator/resource/generate/components/daemonsets_test.go [hunk 3]
```diff
@@ -16,6 +16,60 @@ var _ = Describe("Handler DaemonSet", func() {
...
 	})
 
+	Context("SecurityContext", func() {
+		It("should NOT set ReadOnlyRootFilesystem on privileged virt-handler container", func() {
+			ds := NewHandlerDaemonSet(config, "", "", "")
+			container := ds.Spec.Template.Spec.Containers[0]
+			Expect(container.Name).To(Equal(VirtHandlerName))
+			Expect(container.SecurityContext).ToNot(BeNil())
+			Expect(container.SecurityContext.Privileged).ToNot(BeNil())
+			Expect(*container.SecurityContext.Privileged).To(BeTrue())
```
Comment (0xFelix) [inline pkg/virt-operator/resource/generate/components/daemonsets_test.go]: IIRC there's a BeEquivalent matcher that should allow to combine these

File: pkg/virt-operator/resource/generate/components/daemonsets_test.go [hunk 4]
```diff
@@ -16,6 +16,60 @@ var _ = Describe("Handler DaemonSet", func() {
...
+			Expect(container.SecurityContext.ReadOnlyRootFilesystem).To(BeNil(),
+				"privileged containers should not set readOnlyRootFilesystem as it provides no security benefit and breaks runtime writes")
+		})
+
+		It("should NOT set ReadOnlyRootFilesystem on privileged virt-launcher init container", func() {
```
Comment (0xFelix) [inline pkg/virt-operator/resource/generate/components/daemonsets_test.go]: `virt-launcher` is not a DaemonSet, so I don't think it does belong here. Is there a better place?

File: pkg/virt-operator/resource/generate/components/deployments.go [hunk 5]
```diff
@@ -219,6 +221,21 @@ func attachProfileVolume(spec *corev1.PodSpec) {
 
 }
 
+func attachTmpVolume(spec *corev1.PodSpec) {
+	volume := corev1.Volume{
+		Name: TmpDirName,
+		VolumeSource: corev1.VolumeSource{
+			EmptyDir: &corev1.EmptyDirVolumeSource{},
```
Comment (0xFelix) [inline pkg/virt-operator/resource/generate/components/deployments.go]: What about defining a size limit like `1Gi`?

File: pkg/virt-operator/resource/generate/components/deployments_test.go [hunk 6]
```diff
@@ -13,4 +15,94 @@ var _ = Describe("Deployments", func() {
...
+			}),
+		)
+
+		It("should have emptyDir /tmp volume on virt-api", func() {
+			d := NewApiServerDeployment(&operatorutil.KubeVirtDeploymentConfig{}, "", "", "")
+			podSpec := d.Spec.Template.Spec
+
+			hasTmpVolume := false
+			for _, vol := range podSpec.Volumes {
+				if vol.Name == TmpDirName {
```
Comment (0xFelix) [inline pkg/virt-operator/resource/generate/components/deployments_test.go]: Can it be expressed as a ginkgo matcher?

File: tests/operator/operator.go [hunk 7]
```diff
@@ -2398,6 +2398,86 @@ var _ = Describe("[sig-operator]Operator", Serial, decorators.SigOperator, func(
 			verifyAggregateLabels(kubevirt.Client(), "true")
 		})
 	})
+
+	Describe("Security", func() {
+		It("all infrastructure deployments should have readOnlyRootFilesystem set to true", func() {
```
Comment (0xFelix) [inline tests/operator/operator.go]: I'd say unit level tests for this setting are sufficient. Functional tests are not necessarily required.

File: pkg/virt-operator/resource/generate/components/daemonsets.go [hunk 8]
```diff
@@ -21,14 +21,15 @@ import (
...
 const (
-	VirtHandlerName = "virt-handler"
-	kubeletPodsPath = util.KubeletRoot + "/pods"
-	runtimesPath    = "/var/run/kubevirt-libvirt-runtimes"
-	PrHelperName    = "pr-helper"
-	prVolumeName    = "pr-helper-socket-vol"
-	devDirVol       = "dev-dir"
-	SidecarShimName = "sidecar-shim"
-	etcMultipath    = "etc-multipath"
+	VirtLauncherName = "virt-launcher"
```
Comment (0xFelix) [inline pkg/virt-operator/resource/generate/components/daemonsets.go]: virt-launcher is not a DaemonSet and does not belong into this file.

File: pkg/virt-operator/resource/generate/components/daemonsets.go (hunk 2 above)
Comment (davmarro) [reply pkg/virt-operator/resource/generate/components/daemonsets.go]: Sure!

File: pkg/virt-operator/resource/generate/components/daemonsets_test.go (hunk 3 above)
Comment (davmarro) [reply pkg/virt-operator/resource/generate/components/daemonsets_test.go]: Yes, updated. Thanks

File: pkg/virt-operator/resource/generate/components/daemonsets.go (hunk 8 above)
Comment (davmarro) [reply pkg/virt-operator/resource/generate/components/daemonsets.go]: Yes, I just realized this was used only for a test. Removed.

File: pkg/virt-operator/resource/generate/components/daemonsets_test.go (hunk 4 above)
Comment (davmarro) [reply pkg/virt-operator/resource/generate/components/daemonsets_test.go]: Yes, I understand the point but this test targets the `virt-launcher` init container embedded in the `virt-handler` DaemonSet pod template (and not the virt-launcher itself). That's why I added the test here.

File: pkg/virt-operator/resource/generate/components/deployments.go (hunk 5 above)
Comment (davmarro) [reply pkg/virt-operator/resource/generate/components/deployments.go]: Yes, agree.

File: pkg/virt-operator/resource/generate/components/deployments_test.go (hunk 6 above)
Comment (davmarro) [reply pkg/virt-operator/resource/generate/components/deployments_test.go]: Yes, much better. Thanks

File: tests/operator/operator.go (hunk 7 above)
Comment (davmarro) [reply tests/operator/operator.go]: Ok.

File: pkg/virt-operator/resource/generate/components/daemonsets_test.go [hunk 9]
```diff
@@ -16,6 +16,56 @@ var _ = Describe("Handler DaemonSet", func() {
...
+			ds := NewHandlerDaemonSet(config, "", "", "")
+			container := ds.Spec.Template.Spec.Containers[0]
+			Expect(container.Name).To(Equal(VirtHandlerName))
+			Expect(container.SecurityContext).ToNot(BeNil())
+			Expect(container.SecurityContext.Privileged).To(HaveValue(BeTrue()))
+			Expect(container.SecurityContext.ReadOnlyRootFilesystem).To(BeNil(),
+				"privileged containers should not set readOnlyRootFilesystem as it provides no security benefit and breaks runtime writes")
+		})
+
+		It("should NOT set ReadOnlyRootFilesystem on the privileged node-labeller init container", func() {
```
Comment (0xFelix) [inline pkg/virt-operator/resource/generate/components/daemonsets_test.go]: nit: Should say virt-launcher init container`, right?

Comment (davmarro) [reply pkg/virt-operator/resource/generate/components/daemonsets_test.go]: Yes, sorry.

**Review by dasionov — APPROVED** [review]

File: pkg/virt-operator/resource/generate/components/deployments.go [hunk 10]
```diff
@@ -331,6 +350,7 @@ func NewApiServerDeployment(config *operatorutil.KubeVirtDeploymentConfig, produ
 	attachCertificateSecret(&deployment.Spec.Template.Spec, VirtApiCertSecretName, "/etc/virt-api/certificates")
 	attachCertificateSecret(&deployment.Spec.Template.Spec, VirtHandlerCertSecretName, "/etc/virt-handler/clientcertificates")
 	attachProfileVolume(&deployment.Spec.Template.Spec)
+	attachTmpVolume(&deployment.Spec.Template.Spec)
```
Comment (xpivarc) [inline pkg/virt-operator/resource/generate/components/deployments.go:353]: Why is this needed?

Comment (davmarro) [reply pkg/virt-operator/resource/generate/components/deployments.go]: Hi @xpivarc, this is needed because the virt-api creates a tmp dir (https://github.com/kubevirt/kubevirt/blob/f88039ae275690fc11701b0cbbee98ef395b1c1a/pkg/virt-api/api.go#L190-L189) during the startup and with readOnlyFileSystem set to true, this command will fail. So an empty dir is mounted at /tmp to preserve the existing behavior.
